### PR TITLE
CA1047: Do not declare protected members in sealed types

### LINKED ISSUE #393: Port FxCop rule CA1047: DoNotDeclareProtectedMembersInSealedTypes
**Title:** Do not declare protected members in sealed types

**Description:**

Types declare protected members so that inheriting types can access or override the member. By definition, sealed types cannot be inherited, which means that protected methods on sealed types cannot be called.

**Dependency:** None

**Notes:**
1. If VB compiler doesn't warn, implement VB analyzer. 2. In any case, implement fixer.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #393

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    /// <summary>
+    /// This rule is not implemented for C# as the compiler warning CS0628 already covers this part.
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs]: Oh wow, VB compiler has no diagnostic for this?

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                if (!symbol.MatchesConfiguredVisibility(context.Options, Rule, context.CancellationToken))
+                {
+                    return;
+                }
+
+                if (symbol.IsProtected() &&
+                    !symbol.IsOverride &&
+                    symbol.ContainingType.IsSealed)
+                {
+                    context.ReportDiagnostic(symbol.CreateDiagnostic(Rule, symbol.ToDisplayString(), symbol.ContainingType.ToDisplayString()));
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs]:
```suggestion
                    context.ReportDiagnostic(symbol.CreateDiagnostic(Rule, symbol.Name, symbol.ContainingType.Name));
```

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs (hunk 1 above)
Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs]: Nope or I failed to have it raising an issue with the given code.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                }
+
+                if (!symbol.IsProtected() ||
+                    symbol.IsOverride ||
+                    !symbol.ContainingType.IsSealed)
+                {
+                    return;
+                }
+
+                if (symbol is IMethodSymbol method && method.IsFinalizer())
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs]: @mavasani I missed the special case in the doc!

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        internal static DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Design,
+            RuleLevel.IdeSuggestion,
+            description: s_localizableDescription,
+            isPortedFxCopRule: true,
+            isDataflowRule: false,
+            isEnabledByDefaultInFxCopAnalyzers: false);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDeclareProtectedMembersInSealedTypes.cs:32]: I'd revert `isEnabledByDefaultInFxCopAnalyzers: false` if RuleLevel is not `RuleLevel.Disabled`, i.e. not disabled in NetAnalyzers package.

**Pull request merged**